aptos-move/dynamic-transaction-composer/src/builder.rs

```diff
@@ -0,0 +1,593 @@
+pub struct AllocatedLocal {
+    op_type: ArgumentOperation,
+    is_parameter: bool,
+    local_idx: u16,
+}
+
+#[derive(Tsify, Clone, Debug, Serialize, Deserialize)]
+#[tsify(into_wasm_abi, from_wasm_abi)]
+pub enum Argument {
+    Raw(Vec<u8>),
+    Signer(u16),
+    Local(AllocatedLocal),
+}
+
+#[derive(Clone, Debug)]
+#[wasm_bindgen]
+struct BuilderCall {
+    type_args: Vec<SignatureToken>,
+    call_idx: FunctionHandleIndex,
+    arguments: Vec<AllocatedLocal>,
+    // Assigned index of returned values in the local pool.
+    returns: Vec<u16>,
+
+    #[cfg(test)]
+    type_tags: Vec<TypeTag>,
+}
+
+#[derive(Clone, Debug)]
+#[wasm_bindgen]
+pub struct TransactionComposer {
+    modules: BTreeMap<ModuleId, CompiledModule>,
+    builder: CompiledScriptBuilder,
+    calls: Vec<BuilderCall>,
+    parameters: Vec<Vec<u8>>,
+    locals_availability: Vec<bool>,
+
+    locals_ty: Vec<SignatureToken>,
+    parameters_ty: Vec<SignatureToken>,
+
+    #[cfg(test)]
+    signer_count: u16,
+}
+
+#[wasm_bindgen]
+impl TransactionComposer {
+    /// Create a builder with one distinct signer available. This should be the default configuration.
+    pub fn single_signer() -> Self {
+        let mut script = empty_script();
+        script.code.code = vec![];
+
+        let builder = CompiledScriptBuilder::new(script);
+        Self {
+            modules: BTreeMap::new(),
+            builder,
+            calls: vec![],
+            parameters: vec![],
+            locals_availability: vec![],
+            locals_ty: vec![],
+            parameters_ty: vec![SignatureToken::Reference(Box::new(SignatureToken::Signer))],
+
+            #[cfg(test)]
+            signer_count: 1,
+        }
+    }
+
+    /// Create a builder with one signer needed for script. This would be needed for multi-agent transaction where multiple signers are present.
+    pub fn multi_signer(signer_count: u16) -> Self {
+        let mut script = empty_script();
+        script.code.code = vec![];
+
+        let builder = CompiledScriptBuilder::new(script);
+        Self {
+            modules: BTreeMap::new(),
+            builder,
+            calls: vec![],
+            parameters: vec![],
+            locals_availability: vec![],
+            locals_ty: vec![],
+            parameters_ty: std::iter::repeat(SignatureToken::Reference(Box::new(
+                SignatureToken::Signer,
+            )))
+            .take(signer_count.into())
+            .collect(),
+
+            #[cfg(test)]
+            signer_count,
+        }
+    }
+
+    /// Consume the builder and generate a serialized script with calls in the builder.
+    pub fn generate_batched_calls(self, with_metadata: bool) -> Result<Vec<u8>, String> {
+        self.generate_batched_calls_impl(with_metadata)
+            .map_err(|e| e.to_string())
+    }
+
+    /// Load up a module from a remote endpoint. Will need to invoke this function prior to the call.
+    pub async fn load_module(
```

Comment:
I'm not against the fetching modules from the remote, but I wonder if we should provide an option for people to provide them (or ABIs) locally, potentially for saving the turnaround time + better service availability.